File: clojure/bob/src/bob.clj
(ns bob
  (:require [clojure.string :as str]))

(defn nothing-said? [s]
  (str/blank? s))

(defn ends-with-question? [s]
  (re-find #"\?$" s))

(defn has-lower-case? [s]
  (re-find #"[a-z]+" s))

(defn has-upper-case? [s]
  (re-find #"[A-Z]+" s))

(defn shouting? [s]
  (and (has-upper-case? s) (not (has-lower-case? s))))

(defn shouting-questions? [s]
  (and (ends-with-question? s) (shouting? s)))

(defn response-for [str]
  (let [s (str/trim str)]
    (cond
      (nothing-said? s) "Fine. Be that way!"
      (shouting-questions? s) "Calm down, I know what I'm doing!"
      (shouting? s) "Whoa, chill out!"
      (ends-with-question? s) "Sure."
      :else "Whatever.")))

File: solutions/ruby/bob/1/bob.rb
class Bob
  def self.hey(remark)
    scrubbed = remark.gsub(/[^\w?]+/, '')
    case scrubbed
    when /^[A-Z]+\?$/
      "Calm down, I know what I'm doing!"
    when /\?$/
      'Sure.'
    when /^[^A-Z]+$/
      'Whatever.'
    when /^[A-Z0-9]+$/
      'Whoa, chill out!'
    when /^\s*$/
      'Fine. Be that way!'
    else
      'Whatever.'
    end
  end
end
